File: Fundamental courses of the major/Math For Computer Science/TAI LIEU KHOA TRUOC/C++ & python/C1 Number theory/Prime factorize and perfect number.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;

bool is_prime(int n) {
    if (n <= 1) return false;
    for (int i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) return false;
    }
    return true;
}

void prime_factorization(int n, vector<int>& p, vector<int>& e) {
    for (int x = 2; x <= sqrt(n); x++) {
        if (is_prime(x)) {
            if (n % x == 0) {
                int count = 0;
                while (n % x == 0) {
                    n /= x;
                    count++;
                }
                p.push_back(x);
                e.push_back(count);
            }
        }
    }
    if (n > 1) {
        p.push_back(n);
        e.push_back(1);
    }
}

int sum_factors(const vector<int>& p, const vector<int>& e) {
    int total_sum = 1;
    for (size_t i = 0; i < p.size(); i++) {
        total_sum *= (pow(p[i], e[i] + 1) - 1) / (p[i] - 1);
    }
    return total_sum;
}

int num_factors(const vector<int>& e) {
    int num_of_factors = 1;
    for (int exp : e) {
        num_of_factors *= (exp + 1);
    }
    return num_of_factors;
}

vector<int> factor_list(int num_of_factors, const vector<int>& p, const vector<int>& e) {
    vector<int> factors = {1};
    vector<int> temp_expo(e.size(), 0);
    for (int i = 1; i < num_of_factors; i++) {
        int factor = 1;
        int j = e.size() - 1;
        while (j >= 0 && temp_expo[j] == e[j]) {
            temp_expo[j] = 0;
            j--;
        }
        if (j >= 0) {
            temp_expo[j]++;
        }
        for (size_t k = 0; k < p.size(); k++) {
            factor *= pow(p[k], temp_expo[k]);
        }
        factors.push_back(factor);
    }
    sort(factors.begin(), factors.end());
    return factors;
}

int prod_of_factors(int n, int num_of_factors) {
    return pow(n, num_of_factors / 2);
}

vector<int> perfect_numbers(int limit) {
    vector<int> perfect_nums;
    for (int p = 2; p < limit; p++) {
        int mersenne_prime = pow(2, p) - 1;
        if (is_prime(mersenne_prime)) {
            int perfect_num = pow(2, p - 1) * mersenne_prime;
            if (perfect_num <= limit) {
                perfect_nums.push_back(perfect_num);
            } else {
                break;
            }
        }
    }
    return perfect_nums;
}

int main() {
    int n;
    cout << "Nhap vao so n: ";
    cin >> n;

    if (is_prime(n)) {
        cout << "La so nguyen to!" << endl;
    } else {
        cout << "Khong phai so nguyen to!" << endl;
    }

    vector<int> p, e;
    prime_factorization(n, p, e);

    cout << "Tich cac thua so nguyen to cua " << n << ": " << n << " = ";
    for (size_t i = 0; i < p.size(); i++) {
        cout << p[i];
        if (e[i] > 1) {
            cout << "^" << e[i];
        }
        if (i < p.size() - 1) {
            cout << ".";
        }
    }
    cout << endl;

    int sum_of_factors = sum_factors(p, e);
    cout << "Tong cac uoc so cua " << n << " la: " << sum_of_factors << endl;

    int num = num_factors(e);
    cout << "So uoc so cua " << n << " la: " << num << endl;

    vector<int> factors = factor_list(num, p, e);
    cout << "Cac uoc so cua " << n << " la: ";
    for (int factor : factors) {
        cout << factor << " ";
    }
    cout << endl;

    int prod = prod_of_factors(n, num);
    cout << "Tich cac uoc so cua " << n << " la: " << prod << endl;

    vector<int> perfect_nums = perfect_numbers(n);
    if (!perfect_nums.empty()) {
        cout << "Cac so hoan hao nho hon hoac bang " << n << " la: ";
        for (int perfect_num : perfect_nums) {
            cout << perfect_num << " ";
        }
        cout << endl;
    } else {
        cout << "Khong co so hoan hao nao nho hon hoac bang " << n << "." << endl;
    }

    return 0;
}

File: Fundamental courses of the major/Math For Computer Science/TAI LIEU KHOA TRUOC/C++ & python/C1 Number theory/Prime factorize and perfect number.py
import math

def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

def prime_factorization(n):
    p = []
    e = []
    for x in range(2, int(math.sqrt(n)) + 1):
        if is_prime(x):
            if n % x == 0:
                count = 0
                while n % x == 0:
                    n //= x
                    count += 1
                p.append(x)
                e.append(count)
    if n > 1:
        p.append(n)
        e.append(1)
    return p, e

def sum_factors(p, e):
    total_sum = 1
    for i in range(len(p)):
        total_sum *= (p[i] ** (e[i] + 1) - 1) // (p[i] - 1)
    return total_sum

def num_factors(e):
    num_of_factors = 1
    for exp in e:
        num_of_factors *= (exp + 1)
    return num_of_factors

def factor_list(num_of_factors, p, e):
    factors = [1]
    temp_expo = [0] * len(e)
    i = 1
    while i < num_of_factors:
        factor = 1
        j = len(e) - 1
        while j >= 0 and temp_expo[j] == e[j]:
            temp_expo[j] = 0
            j -= 1
        if j >= 0:
            temp_expo[j] += 1
        for k in range(len(p)):
            factor *= (p[k] ** temp_expo[k])
        factors.append(factor)
        i += 1
    return factors

def prod_of_factors(n, num_of_factors):
    return n ** (num_of_factors // 2)

def perfect_numbers(limit):
    perfect_nums = []
    for p in range(2, limit):
        mersenne_prime = 2 ** p - 1
        if is_prime(mersenne_prime):
            perfect_num = (2 ** (p - 1)) * mersenne_prime
            if perfect_num <= limit:
                perfect_nums.append(perfect_num)
            else:
                break
    return perfect_nums

def main():
    n = int(input("Nhập vào số n: "))
    if is_prime(n):
        print("Là số nguyên tố!")
    else:
        print("Không phải số nguyên tố!")
    p, e = prime_factorization(n)
    num = num_factors(e)
    print(f"Tích các thừa số nguyên tố của {n}:")
    print(f"{n} = ", end="")
    for i in range(len(p)):
        print(f"{p[i]}", end="")
        if e[i] > 1:
            print(f"^{e[i]}", end="")
        if i < len(p) - 1:
            print(".", end="")
    print()
    sum_of_factors = sum_factors(p, e)
    print(f"Tổng các ước số của {n} là: {sum_of_factors}")
    print(f"Số ước số của {n} là: {num}")
    factors = factor_list(num, p, e)
    print(f"Các ước số của {n} là: {' '.join(map(str, factors))}")
    prod = prod_of_factors(n, num)
    print(f"Tích các ước số của {n} là: {prod}")
    perfect_nums = perfect_numbers(n)
    if perfect_nums:
        print(f"Các số hoàn hảo nhỏ hơn hoặc bằng {n} là: {', '.join(map(str, perfect_nums))}")
    else:
        print(f"Không có số hoàn hảo nào nhỏ hơn hoặc bằng {n}.")
main()
